Fixed #30137 -- Replaced OSError aliases with the canonical OSError.

jdufresne opened the pull request:
Starting with Python 3.3, EnvironmentError, IOError, WindowsError,
socket.error, select.error and mmap.error are aliases of OSError.

For additional details, see the Python 3.3 release notes:

https://docs.python.org/3/whatsnew/3.3.html#pep-3151-reworking-the-os-and-io-exception-hierarchy

> You don’t have to worry anymore about choosing the appropriate
> exception type between OSError, IOError, EnvironmentError,
> WindowsError, mmap.error, socket.error or select.error. All these
> exception types are now only one: OSError. The other names are kept as
> aliases for compatibility reasons.

https://code.djangoproject.com/ticket/30137

pope1ni commented on django/contrib/gis/gdal/libgdal.py:
```diff
@@ -13,7 +13,7 @@
     lib_path = settings.GDAL_LIBRARY_PATH
-except (AttributeError, EnvironmentError,
+except (AttributeError, OSError,
         ImportError, ImproperlyConfigured):
```
Collapse these onto one line and order alphabetically.

pope1ni commented on django/contrib/gis/geos/libgeos.py:
```diff
@@ -23,7 +23,7 @@ def load_geos():
         lib_path = settings.GEOS_LIBRARY_PATH
-    except (AttributeError, EnvironmentError,
+    except (AttributeError, OSError,
             ImportError, ImproperlyConfigured):
```
Collapse these onto one line and order alphabetically.

pope1ni commented on django/core/servers/basehttp.py:
```diff
@@ -171,7 +171,7 @@ def handle(self):
         try:
             self.connection.shutdown(socket.SHUT_WR)
-        except (socket.error, AttributeError):
+        except (OSError, AttributeError):
```
Please order alphabetically.

pope1ni commented on django/core/servers/basehttp.py:
```diff
@@ -51,8 +51,8 @@ def get_internal_wsgi_application():
-    exc_type, exc_value = sys.exc_info()[:2]
-    return issubclass(exc_type, socket.error) and exc_value.args[0] == 32
+    exc_type, _exc_value, _traceback = sys.exc_info()
+    return issubclass(exc_type, BrokenPipeError)
```
May as well flatten to one line and use `sys.exc_info()[0]`.

jdufresne replied:
It is minor and probably comes down to personal preference, but I find the current form slightly more readable and less dense. It shows exactly what `sys.exc_info()` returns with descriptive variables names (a new contributor might not be familiar with this function). It is immediately clear that is is a 3-tuple and not something else like a list. This form also desn't overload a single line with so many operations.

But, again, I recognize this is personal preference so this line change is required for it to be agreeable, I'll change it.

jdufresne commented:
Thanks for the review Nick! I've updated all `except` statements to order the exception classes in alphabetical order.

felixxm commented on tests/mail/tests.py:
```diff
@@ -1570,7 +1569,7 @@ def test_fail_silently_on_connection_error(self):
         A socket connection error is silenced with fail_silently=True.
         """
-        with self.assertRaises(socket.error):
+        with self.assertRaises(OSError):
```
I think we can use a more specific subclass i.e. `ConnectionError`.

jdufresne commented:
Thanks for the review :slightly_smiling_face: . I've made the following changes:

- Replaced double quotes with single quotes
- Updated "socket error" comment to reference `OSError`
- `exc_type, _exc_value, _traceback` -> `exc_type, _, _`
- Wrapped comments to 79
- Used more specific subclasses where appropriate

jdufresne commented:
Sorry @timgraham I didn't realize you were working on this simultaneously. I may have force pushed over some of your edits.

timgraham commented on tests/template_tests/test_loaders.py:
```diff
@@ -191,11 +191,11 @@ def test_permissions_error(self):
 
     def test_notafile_error(self):
-        with self.assertRaises(IOError):
+        with self.assertRaises(IsADirectoryError):
```
This isn't working on Windows:
```
======================================================================
ERROR: test_notafile_error (template_tests.test_loaders.FileSystemLoaderTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\tim\code\django\tests\template_tests\test_loaders.py", line 199, in test_notafile_error
    self.engine.get_template('first')
  File "c:\users\tim\code\django\django\template\engine.py", line 143, in get_template
    template, origin = self.find_template(template_name)
  File "c:\users\tim\code\django\django\template\engine.py", line 125, in find_template
    template = loader.get_template(name, skip=skip)
  File "c:\users\tim\code\django\django\template\loaders\base.py", line 24, in get_template
    contents = self.get_contents(origin)
  File "c:\users\tim\code\django\django\template\loaders\filesystem.py", line 23, in get_contents
    with open(origin.name, encoding=self.engine.file_charset) as fp:
PermissionError: [Errno 13] Permission denied: 'C:\\Users\\tim\\code\\django\\tests\\template_tests\\templates\\first'
```
It seems that `with self.assertRaises(IOError):` also catches `PermissionError`.

Should we go with `PermissionError if sys.platform.startswith('win') else IsADirectoryError` here?

jdufresne replied:
That works for me. Would you like me to open a PR with that fix?

I'd also be OK catching the parent `OSError` if you think the platform specificity isn't useful in this case.
